Added ReadableStream, WritableStream and DuplexStream

utvara opened the pull request:
Purpose of this PR is to add node stream interfaces to 0mq socket. Motivation for this is that in case you have high frequency inbound messages there is no reliable way of preventing memory leaks. Thou my opinion is that we should decouple socket handling object form eventemitter, I did not tackle that problem at the moment. 

I've added a new classes:
- ReadableStream - extension of Readable stream that accepts pull, sub, xsub Socket and provides socket API on top of it
- WritableStream - extension of Writable stream that accepts push, pub, xpub Socket and provides socket API on top of it
- DuplexStream - extension of Duplex stream that accepts req, xreq, rep, xrep, dealer, router, pair Socket and provides socket API on top of it

For next major version I think we should consider separating the Socket class and having clean interfaces using:
- EventEmmiter
- Stream

NOTE: I am forcing the use of objectMode on readable stream to support multipart messages.

Please leave your notes and comments.

reqshark commented on lib/index.js:
```diff
@@ -507,14 +527,13 @@ Socket.prototype._flush = function() {
     if (flags & zmq.ZMQ_POLLIN) {
-      emitArgs = ['message'];
+      emitArgs = this._read();
 
```
do you mean `this.read()`? i'm not sure we should be calling `_read()` directly.

utvara replied:
I've just extracted the reading part from _flush into _read. We could make it "public" and rename it to **read**

reqshark replied:
if behavior is the same it doesn't matter, i was just quoting [the docs](https://iojs.org/api/stream.html#stream_readable_read_size_1)

utvara replied:
Socket is not related to/inheriting from stream, check the implementation of ReadableStream for that.

reqshark commented:
I think Object mode makes sense since otherwise your default high water mark is like 16K messages

reqshark commented:
> Thou my opinion is that we should decouple socket handling object form event emmiter, I did not tackle that problem at the moment.

+1

reqshark commented:
woah this is cool, question: where is the `Socket` or `socket` exactly inheriting from Readable?

Does that just happen from returning a new readable off the constructor call in `createReadableSocket`?

utvara commented:
Socket is not inheriting from Readable. ReadableStream implementation accepts ReadbaleSocket as first parameter
```
var ReadableStream = function (socket, options) {
```

so usage would be:

```
var socket = zmq.createReadableSocket('pull');
socket.connect(port);

var stream = zmq.createReadableStream(socket, {highWaterMark: 100});
```
I didn't want to use multiple inheritance or mixin, so you need to create the socket and provide it to the ReadableStream. 

This could also be handled by constructor but it would require socket type, socket options and stream options. Plus you still need access to socket object if we don't put a fasade in stream object.

reqshark commented:
btw node 0.8 is going to fail tests all day unless these streams are subclassed from [`readable-stream`](https://github.com/iojs/readable-stream)

maybe we ought to not worry about another dep, instead just drop support for node 0.8?

utvara commented:
yes, for me Stream1 is a no go. I would prefer to always use Stream3. It is an open question what to do. I could make the new interfaces dependant to node 0.10+ so that overall dependency doesn't change. Same way we handle multiple 0mq versions...

reqshark commented:
I believe that if the stream's internal `_read(n)` function doesn't moderate inbound message I/O, there's no backpressure mechanism available

utvara commented:
It depends how you use the stream, If you are using it as in the example then you are controlling the pull in speed.
```
var stream = zmq.createReadableStream(socket, {highWaterMark: 100});
setInterval(function () {
  var message;
  for (var i = 0; i < 10; i += 1) {
     message = stream.read();

     if (!message) return;
     console.log(socket.identity + ': received data ', message.toString());
   }
}, 1000);
```
this will create a stream with internal buffer of 100 objects, and pull additional 10 objects from 0mq every second. 

However if you were to use stream .on 'data' event you will not be able to have that precise control. You will still be able to pause the stream.

reqshark commented:
it's just a random stream queuing random messages and when it's high water mark is reached: 

https://github.com/iojs/io.js/blob/v1.x/lib/_stream_readable.js#L171

utvara commented:
I've tested with HWM and number of reads from 0mq is fully controlled with number of stream read invocations.

reqshark commented:
with multiple pull sockets connected?

utvara commented:
Yes, I've tested with one push and 3 pull sockets. I was counting how many times stream **push** method was called.

utvara commented:
@kkoopa @ronkorving any opinion on stream support...

ronkorving commented on lib/index.js:
```diff
@@ -479,6 +479,26 @@ Socket.prototype.send = function(msg, flags) {
+    } while (this._zmq.getsockopt(zmq.ZMQ_RCVMORE));
+
+    return message;
+  } else {
```
No need for `else` after a return statement.

ronkorving commented:
Is there really a need for a `createReadableSocket`? Why can't I just pass my socket as I would normally have created it? You could always overwrite or augment `onReady` from the readable stream.

dominicumbeer commented on lib/readable_stream.js:
```diff
@@ -0,0 +1,54 @@
+  this._readMore = true;
+};
+
+ReadableStream.prototype._read = function() {
```
This function isn't used, right?

utvara commented:
OK, I've removed the ReadableSocket and added autoFlush option to Socket. Setting autoFlush to false prevents _flush to be automatically executed. Also ReadableStream now accepts Socket with specific type and autoFlush turned off. I'm looking into how to extract write method from _flush.

ronkorving commented on lib/readable_stream.js:
```diff
@@ -0,0 +1,58 @@
+    }
+  }
+
+  // We would like to read more, but there isn't any.
```
Is that even possible? I thought ZMQ guarantees atomicity?

utvara replied:
readMore refers to socket wants to read more, but we didn't get any more messages from 0mq. Basically it means that as soon as there is ready event we should push the message to socket and not wait for it to call _read

utvara commented:
I've added WritableStream as well.

utvara commented:
I've added DuplexStream as well, now all socket types apart from zmq_stream (don't know how it works) are supported. There is now just one constructor function to rule them all **createStream** that accepts socket and stream options. I would add more tests to PR if you think it makes sense to continue with this.  Note that this PR makes https://github.com/JustinTulloss/zeromq.node/pull/402 obsolete.

reqshark commented:
hmm ...would could definitely stream ZMQ_STREAM socket type

ronkorving commented on lib/index.js:
```diff
@@ -479,6 +488,86 @@ Socket.prototype.send = function(msg, flags) {
+    server.once('ready', function () {
+      self.write(parts, callback);
+    });
+    return;
```
`return this;`? like on line 568

ronkorving commented on lib/index.js:
```diff
@@ -479,6 +488,86 @@ Socket.prototype.send = function(msg, flags) {
+      // This loop will pull off the items up until and including
+      // the first item that is not flagged SNDMORE.
+
+      if (callback) return callback(sendError);
```
You're not returning `this` here either.

ronkorving replied:
If there is no callback, should we not return?

utvara replied:
problem is that this can be delayed till socket is ready, then there is no point in returning/throwing an error.

ronkorving commented on lib/index.js:
```diff
@@ -479,6 +488,86 @@ Socket.prototype.send = function(msg, flags) {
+    } catch (sendError) {
+      // More chunks were to follow, which we should now drop.
+      // This loop will pull off the items up until and including
+      // the first item that is not flagged SNDMORE.
```
What happened to our loop?

utvara replied:
not needed anymore, write accepts String|Buffer|Array and doesn't accept flag param, so we can just drop the current multipart message

ronkorving commented on lib/stream.js:
```diff
@@ -0,0 +1,154 @@
+  , Writable = require('stream').Writable
+  , Duplex = require('stream').Duplex;
+
+var ReadableStream = function (socket, options) {
```
Mind using declaration syntax? Ie: `function ReadableStream (socket, options) {`. Makes debugging a bit easier.

ronkorving commented on lib/stream.js:
```diff
@@ -0,0 +1,154 @@
+      socket.type != 'xrep' &&
+      socket.type != 'dealer' &&
+      socket.type != 'router' &&
+      socket.type != 'pair') {
```
Please use `!==` consistently.

ronkorving commented on lib/index.js:
```diff
@@ -479,6 +488,83 @@ Socket.prototype.send = function(msg, flags) {
+  if (!(flags & zmq.ZMQ_POLLOUT)) {
+    server.once('ready', function () {
+      self.write(parts, callback);
+    });
```
This really scares me under high load. I fear this could really cause a big memory spike (a closure here and inside EventEmitter for "once" iirc).

utvara replied:
Sure, we on our end broke the _outgoing regularly :). This is not to called lightly. In writable stream case buffering is happening in stream.Writable class and it is calling **write** in sequence mode, we would need a pref test to see direct impact. 

I'm using 0mq with 100k messages per sec in a distributed system, we have had mem leaks on push and pull side, so I would say we managed to break both ends. More specific **_outgoing** and eventEmitter are the sources of leak. 

We already have this PR in production and it is solving the problem on pull side by piling up data in 0mq and not in heap. Potential downside that I saw, but I can not guaranty, is that it seems to me that system load went up due to constant switching between node and 0mq...

We did not use Writable side in production yet.

ronkorving replied:
So perhaps instead of depending on the event emitter, can we keep the parts on a temporary variable `queuedWriteParts` or something, and have the onReady function call `this.write(this.queuedWriteParts);` ... meh, that's ugly too and we would also need to keep the callback around.

Any suggestions that would allow us to not use EventEmitter here and ideally also not create a closure?

utvara replied:
Callback is needed for the writable stream. So you will either pile up closures or you pile up data in arrays (streams in object mode would do instead of _outgoing). 

Bottom line there is no way that we can take responsibility from developer if there is a mem leak due to high injection rate. 

PS.> In any other language this would block according to 0mq spec.

utvara commented:
@ronkorving @kkoopa @reqshark I've been running some tests with Writable stream, and it seems that my solution is not working if socket is not ready. I was under the impression that onReady is fired if socket changes state to ZMQ_POLLOUT or ZMQ_POLLIN...
